### ISSUE TITLE
[Spark 1.4] ClassCastException: java.lang.String cannot be cast to org.apache.spark.sql.types.UTF8String

### ISSUE THREAD

**Issue by evacchi** [databricks/spark-csv#64]
Currently broken with Spark 1.4 when an explicit schema with non-string fields is specified; e.g.:

```scala
val schema = StructType(Seq(StructField("id", LongType, false)))
```

**Comment by falaki** [comment]
Please provide additional details including sample of data and the code that leads to this error.

**Comment by rxin** [comment]
This is probably because we don't do inbound conversion in Spark itself. cc @yhuai

**Comment by pwendell** [comment]
@yhuai do you mind creating a JIRA to track this - it could be release blocker (/cc @marmbrus)

**Comment by yhuai** [comment]
I see. The issue is that Spark-CSV is using the Spark SQL's internal API to do the type cast, which assumes values are stored in Spark SQL's internal representation (e.g. `UTF8String`). We will fix it in Spark-CSV.


btw, here is the stack trace.

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 31.0 failed 4 times, most recent failure: Lost task 0.3 in stage 31.0 (TID 41, 192.168.1.26): java.lang.ClassCastException: java.lang.String cannot be cast to org.apache.spark.sql.types.UTF8String
	at org.apache.spark.sql.catalyst.expressions.Cast$$anonfun$castToLong$1$$anonfun$apply$28.apply(Cast.scala:249)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$buildCast(Cast.scala:111)
	at org.apache.spark.sql.catalyst.expressions.Cast$$anonfun$castToLong$1.apply(Cast.scala:249)
	at org.apache.spark.sql.catalyst.expressions.Cast.eval(Cast.scala:436)
	at org.apache.spark.sql.catalyst.expressions.InterpretedMutableProjection.apply(Projection.scala:68)
	at org.apache.spark.sql.catalyst.expressions.InterpretedMutableProjection.apply(Projection.scala:52)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:150)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:130)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$class.isEmpty(Iterator.scala:256)
	at scala.collection.AbstractIterator.isEmpty(Iterator.scala:1157)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1.apply(ExistingRDD.scala:63)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1.apply(ExistingRDD.scala:62)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$17.apply(RDD.scala:686)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$17.apply(RDD.scala:686)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
	at org.apache.spark.scheduler.Task.run(Task.scala:70)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by saifellafi** [comment]
Hello, 

I am finding myself in a situation, where data coming from spark-csv fails to be casted. I am trying to get the data into a PairedRRD:

            sqlc.read.format("com.databricks.spark.csv").
            option("header", "true").
            option("charset", "UTF8").
            load("/path/to/file.csv").

            map[(BigInt, Double)](pck_col => (
                pck_col.getAs[BigInt](0),
                pck_col.getAs[Double](13))
            )

When I try to apply any PairedRRD action (lookup, take, etc..) I get a Cast exception.
Does this problem apply to my case?

**Comment by yhuai** [comment]
@saifellafi What is your exception?

**Comment by saifellafi** [comment]
Running from a zeppelin notebook:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 644.0 failed 1 times, most recent failure: Lost task 0.0 in stage 644.0 (TID 2105, localhost): java.lang.ClassCastException Driver stacktrace: at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1266) at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1257) at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1256) at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59) at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47) at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1256) at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:730) at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:730) at scala.Option.foreach(Option.scala:236) at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:730) at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1450) at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1411) at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)

**Comment by yhuai** [comment]
@saifellafi Seems it is not the complete stack trace. I think you can find the real exception appearing below this message. Can you add that?

**Comment by saifellafi** [comment]
Appreciate your time, and sorry for bad output: allow me throw this:

org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 3.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3.0 (TID 5, localhost): java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Double at scala.runtime.BoxesRunTime.unboxToDouble(BoxesRunTime.java:119) at scala.runtime.ScalaRunTime$.array_update(ScalaRunTime.scala:91) at scala.Array$.slowcopy(Array.scala:81) at scala.Array$.copy(Array.scala:107) at scala.collection.mutable.ResizableArray$class.copyToArray(ResizableArray.scala:77) at scala.collection.mutable.ArrayBuffer.copyToArray(ArrayBuffer.scala:47) at scala.collection.TraversableOnce$class.copyToArray(TraversableOnce.scala:241) at scala.collection.AbstractTraversable.copyToArray(Traversable.scala:105) at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:249) at scala.collection.AbstractTraversable.toArray(Traversable.scala:105) at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:252) at scala.collection.AbstractIterator.toArray(Iterator.scala:1157) at org.apache.spark.rdd.RDD$$anonfun$take$1$$anonfun$28.apply(RDD.scala:1272) at org.apache.spark.rdd.RDD$$anonfun$take$1$$anonfun$28.apply(RDD.scala:1272) at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765) at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765) at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63) at org.apache.spark.scheduler.Task.run(Task.scala:70) at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) at java.lang.Thread.run(Thread.java:745) Driver stacktrace: at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1266) at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1257) at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1256) at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59) at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47) at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1256) at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:730) at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:730) at scala.Option.foreach(Option.scala:236) at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:730) at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1450) at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1411) at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)

**Comment by yhuai** [comment]
@saifellafi oh i see. For now, if you do not specify the data types of columns, spark-csv will use `StringType` as the type of columns. Assuming your are using scala, you can check the schema of the dataframe by using, 
```
val df = sqlc.read.format("com.databricks.spark.csv").
        option("header", "true").
        option("charset", "UTF8").
        load("/path/to/file.csv")
df.printSchema()
```
You should see types of columns are all `string`. Actually, since your file has a header, you can use
```
df.select($"nameOfCol0".cast("bigint"), $"nameOfCol13".cast("double"))
```
to get a dataframe with the two columns you want.

btw, if you are using Spark 1.4, make sure you use spark-csv 1.1.0 version.

**Comment by saifellafi** [comment]
I can't thank you enough. Been stuck with that issue for two days. I am learning both scala and spark, definitely need to refresh some concepts

**Comment by yhuai** [comment]
@saifellafi Glad I can help! If you have any further question related to spark and spark csv (or other spark libs), feel free to send your question to spark user mailing list user@spark.apache.org.

**Comment by 1stdayonthejob** [comment]
I'm using pyspark and explicitly specifying the schema but still throws when I try to do .show() or .rdd

Input:
~~~
start,stop
3,50
16,200
~~~

~~~python
tab2_df = (sqlContext.read.format('com.databricks.spark.csv')
           .options(header='true', delimiter=",")
           .schema(
               types.StructType([
                   types.StructField('start', types.IntegerType(), False),
                   types.StructField('stop', types.IntegerType(), False)
               ])
           )
           .load('tab2.tsv'))
tab2_df.printSchema()
~~~

schema:
~~~
root
 |-- start: integer (nullable = false)
 |-- stop: integer (nullable = false)
~~~

stacktrace:
~~~
ERROR CsvRelation$: Exception while parsing line: 3,50.
java.lang.ClassCastException: java.lang.String cannot be cast to org.apache.spark.sql.types.UTF8String
	at org.apache.spark.sql.catalyst.expressions.Cast$$anonfun$castToInt$1$$anonfun$apply$32.apply(Cast.scala:265)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$buildCast(Cast.scala:111)
	at org.apache.spark.sql.catalyst.expressions.Cast$$anonfun$castToInt$1.apply(Cast.scala:265)
	at org.apache.spark.sql.catalyst.expressions.Cast.eval(Cast.scala:436)
	at org.apache.spark.sql.catalyst.expressions.InterpretedMutableProjection.apply(Projection.scala:68)
	at org.apache.spark.sql.catalyst.expressions.InterpretedMutableProjection.apply(Projection.scala:52)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:150)
	at com.databricks.spark.csv.CsvRelation$$anonfun$com$databricks$spark$csv$CsvRelation$$parseCSV$1.apply(CsvRelation.scala:130)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:308)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:48)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:103)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:47)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:273)
	at scala.collection.AbstractIterator.to(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:265)
	at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:252)
	at scala.collection.AbstractIterator.toArray(Iterator.scala:1157)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$3.apply(SparkPlan.scala:143)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$3.apply(SparkPlan.scala:143)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
	at org.apache.spark.scheduler.Task.run(Task.scala:70)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
~~~

I've tried callling select on the result DataFrame using 'int' and 'string' and both throws the same exceptions

**Comment by yhuai** [comment]
@1stdayonthejob What is the version of Spark CSV? Can you try 1.1.0 (this issue should be fixed in this version)?

**Comment by snudurupati** [comment]
Just want to confirm that I was having a similar issue with Spark 1.4.0 and Spark CSV 1.0.3, upgrading to Spark CSV 1.1.0 resolved the issue.

**Comment by 1stdayonthejob** [comment]
@yhuai yes you're right it works now, must've been my copy and pasting from the readme and I see that it's been updated now.

Thanks!

**Comment by yhuai** [comment]
@falaki @evacchi I am going to close this issue since it has been fixed in Spark CSV 1.1.0.

**Comment by crakjie** [comment]
I got this exception with spark 1.4.1 and CSV 1.3.0 with spark cassandra connector 1.4.3

Is it a regression  ?
